### PR TITLE
add preprocessing mlp for hstu

### LINKED ISSUE #97: [FEA] Decouple embedding dim and hstu input dim
**Is your feature request related to a problem? Please describe.**
Add a MLP to convert embedding output to the hstu required input dim.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**
Add any other context, code examples, or references to existing implementations about the feature request here.

-----
By submitting this issue, you agree to follow our [code of conduct](https://docs.rapids.ai/resources/conduct/) and our [contributing guidelines](https://github.com/jarmak-nv/rapids-repo-template/blob/main/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by shijieliu**
## Description
resolve #97 

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rapids-repo-template/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: examples/hstu/modules/hstu_block.py [hunk 1]
```diff
@@ -39,6 +37,24 @@ def __init__(
         if self.config.fp16:
             self._training_dtype = torch.float16
 
+        self._item_mlp = None
+        self._contextual_mlp = None
+        if config.hstu_preprocessing_config is not None:
+            self._item_mlp = MLP(
+                in_size=config.hstu_preprocessing_config.item_embedding_dim,
+                layer_sizes=[config.hidden_size, config.hidden_size],
+                activation="relu",
```
Comment (JacoCheung) [inline examples/hstu/modules/hstu_block.py]: How do you think of making `layer_sizes` , `activation` and `bias` configurable? @shijieliu

File: examples/hstu/modules/hstu_block_inference.py [hunk 2]
```diff
@@ -55,61 +39,6 @@ def __init__(
...
-            JaggedData: The preprocessed jagged data, ready for further processing in the HSTU architecture.
-        """
-
-        # Interleaving & Concatenation
-        jd = hstu_preprocess_embeddings(
-            embeddings, batch, is_inference=True, dtype=self._dtype
-        )
-        device = jd.seqlen_offsets.device
-        jd.num_candidates = jd.num_candidates.to(device=device)
-        jd.num_candidates_offsets = jd.num_candidates_offsets.to(device=device)
```
Comment (shijieliu) [inline examples/hstu/modules/hstu_block_inference.py:83]: why do we need these? @geoffreyQiu

File: examples/hstu/modules/hstu_processor.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+        self._item_mlp = None
+        self._contextual_mlp = None
+        if config.hstu_preprocessing_config is not None:
+            self._item_mlp = MLP(
+                in_size=config.hstu_preprocessing_config.item_embedding_dim,
+                layer_sizes=[config.hidden_size, config.hidden_size],
+                activation="relu",
+                bias=True,
+            )
+            self._contextual_mlp = MLP(
```
Comment (JacoCheung) [inline examples/hstu/modules/hstu_processor.py]: What if there is no contextual feature? ( i.e. the contextual embedding is not present like [our benchmark setting](https://github.com/NVIDIA/recsys-examples/blob/9a556c8f88ff5ed97d0d83c40b65ec3e2863df0b/examples/hstu/benchmark_ranking.gin#L37)

File: examples/hstu/training/utils.py [hunk 4]
```diff
@@ -71,6 +72,16 @@ def create_hstu_config(network_args: NetworkArgs):
         num_time_buckets=2048,
         use_time_encoding=False,
     )
+    if (
+        network_args.item_embedding_dim > 0
+        and network_args.contextual_embedding_dim > 0
```
Comment (JacoCheung) [inline examples/hstu/training/utils.py]: item & contextual shuold not be tied.

File: examples/hstu/modules/hstu_block_inference.py [hunk 5]
```diff
@@ -55,61 +39,6 @@ def __init__(
...
-
-        Args:
-            jd (JaggedData): The jagged data output from the HSTU architecture that needs further processing.
-
-        Returns:
-            JaggedData: The postprocessed jagged data.
-        """
-
-        return hstu_postprocess_embeddings(jd, is_inference=True)
-
```
Comment (geoffreyQiu) [inline examples/hstu/modules/hstu_block_inference.py:112]: Need to change the [preprocess](https://github.com/NVIDIA/recsys-examples/blob/main/examples/hstu/model/inference_ranking_gr.py#L361) and [postprocess](https://github.com/NVIDIA/recsys-examples/blob/main/examples/hstu/model/inference_ranking_gr.py#L403) in inference as well

File: examples/hstu/modules/hstu_processor.py [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+
+    if (
+        batch.num_candidates is not None
+        and batch.action_feature_name is not None
+        and not is_inference
+    ):
+        num_candidates = batch.num_candidates * 2
+        max_num_candidates = batch.max_num_candidates * 2
+    else:
+        num_candidates = batch.num_candidates
```
Comment (geoffreyQiu) [inline examples/hstu/modules/hstu_processor.py:127]:
```suggestion
        num_candidates = batch.num_candidates.cuda()
```
In inference, we accept `batch` with values on CPU.

File: examples/hstu/modules/hstu_block_inference.py [hunk 7]
```diff
@@ -125,10 +54,11 @@ def forward(
         Returns:
             JaggedData: The output jagged data.
         """
-        jd = self.hstu_preprocess(embeddings, batch)
-        for hstu_layer in self._attention_layers:
-            jd = hstu_layer(jd)
-        return self.hstu_postprocess(jd)
+        with torch.inference_mode():
```
Comment (geoffreyQiu) [inline examples/hstu/modules/hstu_block_inference.py:57]: This forward function is not directly used in inference.
Need to change the [preprocess](https://github.com/NVIDIA/recsys-examples/blob/main/examples/hstu/model/inference_ranking_gr.py#L361) and [postprocess](https://github.com/NVIDIA/recsys-examples/blob/main/examples/hstu/model/inference_ranking_gr.py#L403) in inference here, as well as the test ([1](https://github.com/NVIDIA/recsys-examples/blob/main/examples/hstu/test/test_hstu_block_inference.py#L145) , [2](https://github.com/NVIDIA/recsys-examples/blob/main/examples/hstu/test/test_hstu_block_inference.py#L210))

File: examples/hstu/modules/hstu_block_inference.py (hunk 2 above)
Comment (geoffreyQiu) [reply examples/hstu/modules/hstu_block_inference.py:83]: This is required for input batch on CPU.

**Comment by shijieliu** [comment]
[ci](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/32375501)

File: examples/hstu/modules/hstu_processor.py (hunk 6 above)
Comment (shijieliu) [reply examples/hstu/modules/hstu_processor.py:127]: fixed

File: examples/hstu/training/utils.py (hunk 4 above)
Comment (shijieliu) [reply examples/hstu/training/utils.py]: fixed

File: examples/hstu/modules/hstu_processor.py (hunk 3 above)
Comment (shijieliu) [reply examples/hstu/modules/hstu_processor.py]: fixed

**Review by geoffreyQiu — APPROVED** [review]
Reviewed the inference related part. 
LGTM.

**Comment by shijieliu** [comment]
[CI](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/32533679)

File: examples/hstu/modules/hstu_processor.py [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+
+        Args:
+            embeddings (Dict[str, JaggedTensor]): A dictionary of embeddings where each key corresponds to a feature name and the value is a jagged tensor.
+            batch (RankingBatch): The batch of ranking data.
+
+        Returns:
+            JaggedData: The preprocessed jagged data, ready for further processing in the HSTU architecture.
+        """
+        device = torch.device("cuda", torch.cuda.current_device())
+        batch = batch.to(device)
```
Comment (JacoCheung) [inline examples/hstu/modules/hstu_processor.py:273]: Why do we need a copy? Shouldn't this be done by data loader?

File: examples/hstu/modules/hstu_processor.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+            jd.values = self._positional_encoder(
+                max_seq_len=jd.max_seqlen,
+                seq_lengths=jd.seqlen,
+                seq_offsets=jd.seqlen_offsets,
+                seq_timestamps=None,
+                seq_embeddings=jd.values,
+                num_targets=jd.num_candidates,
+            )
+
+        if not self._is_inference:
```
Comment (JacoCheung) [inline examples/hstu/modules/hstu_processor.py]: I think we can simply use
```
jd.values = torch.nn.functional.dropout(
                jd.values,
                p=self._dropout_ratio,
                training=not self._is_inference,
            ).to(self._training_dtype)
```

Comment (shijieliu) [reply examples/hstu/modules/hstu_processor.py]: I dont think so.

the traing is deciding whether it's train or eval. The is_inference is controling whether it's inference.

File: examples/hstu/modules/hstu_processor.py (hunk 8 above)
Comment (shijieliu) [reply examples/hstu/modules/hstu_processor.py:273]: The inference dataset will generate cpu data right now.

**Review by JacoCheung — APPROVED** [review]
